### PR TITLE
ENH: stack: allow naming of measure/value column names

### PR DESCRIPTION
**Pull request by kdmurray91**
In R's `reshape2::melt`  it is possible to use a custom name for the newly created stacked columns. This adds two optional keyword arguments that allow this behaviour (demonstrated below).

I have added tests, but I would like to see if I can simplify the diff. I'd like to avoid having to define these kwargs for all the various `stack` and `melt` invocations (suggestions welcome).

## TODO:

- [x] ~Decide on final name for keyword arguments: `variable_name` vs `measure_name`.~ Decision: `variable_name`

```julia
julia> d = DataFrame(red=rand(2), blue=rand(2), green=rand(2), sample=["A", "B"])
2×4 DataFrames.DataFrame
│ Row │ red      │ blue     │ green    │ sample │
├─────┼──────────┼──────────┼──────────┼────────┤
│ 1   │ 0.134251 │ 0.668584 │ 0.463817 │ A      │
│ 2   │ 0.784527 │ 0.46851  │ 0.634924 │ B      │

julia> ds = stack(d, [:red, :blue, :green])
6×3 DataFrames.DataFrame
│ Row │ variable │ value    │ sample │
├─────┼──────────┼──────────┼────────┤
│ 1   │ red      │ 0.134251 │ A      │
│ 2   │ red      │ 0.784527 │ B      │
│ 3   │ blue     │ 0.668584 │ A      │
│ 4   │ blue     │ 0.46851  │ B      │
│ 5   │ green    │ 0.463817 │ A      │
│ 6   │ green    │ 0.634924 │ B      │

julia> dsn = stack(d, [:red, :blue, :green], variable_name=:colour,
                   value_name=:mean_pixel_value)
6×3 DataFrames.DataFrame
│ Row │ colour │ mean_pixel_value │ sample │
├─────┼────────┼──────────────────┼────────┤
│ 1   │ red    │ 0.134251         │ A      │
│ 2   │ red    │ 0.784527         │ B      │
│ 3   │ blue   │ 0.668584         │ A      │
│ 4   │ blue   │ 0.46851          │ B      │
│ 5   │ green  │ 0.463817         │ A      │
│ 6   │ green  │ 0.634924         │ B      │
```

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -38,9 +38,14 @@ melt(df::AbstractDataFrame, id_vars)
   stacking, a normal column indexing type; for `stack` defaults to all
   variables that are not `measure_vars`
 
+
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Unrelated.

File: src/abstractdataframe/reshape.jl [hunk 2]
```diff
@@ -68,43 +73,49 @@ d1m = melt(d1, [:a, :b, :e])
 ```
 
 """
-function stack(df::AbstractDataFrame, measure_vars::Vector{Int}, id_vars::Vector{Int})
+function stack(df::AbstractDataFrame, measure_vars::Vector{Int}, id_vars::Vector{Int}; measure_name=:variable, value_name=:value)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Please break lines so as not to exceed 92 chars.

File: src/abstractdataframe/reshape.jl [hunk 3]
```diff
@@ -18,11 +18,11 @@ Stacks a DataFrame; convert from a wide to long format
 
 
 ```julia
-stack(df::AbstractDataFrame, measure_vars, id_vars)
-stack(df::AbstractDataFrame, measure_vars)
-stack(df::AbstractDataFrame)
-melt(df::AbstractDataFrame, id_vars, measure_vars)
-melt(df::AbstractDataFrame, id_vars)
+stack(df::AbstractDataFrame, measure_vars, id_vars; measure_name=:variable, value_name=:value)
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: `melt` from data.table calls the first argument `variable_name`. I wonder which one is best. (AFAICT only `value_name` exists in reshape2.)

File: src/abstractdataframe/reshape.jl [hunk 4]
```diff
@@ -38,9 +38,14 @@ melt(df::AbstractDataFrame, id_vars)
   stacking, a normal column indexing type; for `stack` defaults to all
   variables that are not `measure_vars`
 
+
 If neither `measure_vars` or `id_vars` are given, `measure_vars`
 defaults to all floating point columns.
 
+The names of the resulting "stacked" columns can be controlled with keyword
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Should use a bullet for each argument. You can move the paragraph above to the `measure_vars` bullet so that it's not out of place.

File: src/abstractdataframe/reshape.jl [hunk 5]
```diff
@@ -68,43 +73,49 @@ d1m = melt(d1, [:a, :b, :e])
...
-    stack(df, mv_inds, _setdiff(1:ncol(df), mv_inds))
+    stack(df, mv_inds, _setdiff(1:ncol(df), mv_inds);
+          measure_name=measure_name, value_name=value_name)
 end
 
 """
 Stacks a DataFrame; convert from a wide to long format; see
 `stack`.
 """
 melt(df::AbstractDataFrame, id_vars::@compat(Union{Int,Symbol})) = melt(df, [id_vars])
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Should add the keyword arguments here too (and below). Also please add tests for them (which are apparently missing else you'd have noticed their lack).

File: src/abstractdataframe/reshape.jl (hunk 3 above)
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: I chose `measure_name` as we use `measure_vars` through this code.  I'm open to change this though.

Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Yeah, but positional argument names can be changed without affecting existing code, which isn't the case for keyword arguments. So now is the time to make a decision. What do others think about it?

**Comment by kdmurray91** [comment]
@nalimilan 1ff78cb should fix most of your concerns.

File: src/abstractdataframe/reshape.jl [hunk 6]
```diff
@@ -65,49 +76,74 @@ d1 = DataFrame(a = repeat([1:3;], inner = [4]),
...
+          measure_name=measure_name, value_name=value_name)
 end
 
 """
 Stacks a DataFrame; convert from a wide to long format; see
 `stack`.
 """
-melt(df::AbstractDataFrame, id_vars::@compat(Union{Int,Symbol})) = melt(df, [id_vars])
-function melt(df::AbstractDataFrame, id_vars)
+function melt(df::AbstractDataFrame, id_vars::@compat(Union{Int,Symbol});
```
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: I felt the concise function definition form was less readable in many cases once the new keyword arguments were added. I'm happy to revert this if others disagree.

File: src/abstractdataframe/reshape.jl [hunk 7]
```diff
@@ -420,40 +456,58 @@ d1m = meltdf(d1, [:a, :b, :e])
...
-    stackdf(df, [measure_var], [id_var])
+function stackdf(df::AbstractDataFrame, measure_var::Int, id_var::Int;
+                 measure_name::Symbol=:variable, value_name::Symbol=:value)
+    stackdf(df, [measure_var], [id_var]; measure_name=measure_name,
+            value_name=value_name)
 end
-function stackdf(df::AbstractDataFrame, measure_vars, id_var::Int)
-    stackdf(df, measure_vars, [id_var])
+function stackdf(df::AbstractDataFrame, measure_vars, id_var::Int;
+                 measure_name::Symbol=:variable, value_name::Symbol=:value)
```
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: Do `measure_name` and `value_name` need this type annotation? I seem to have inconsistently used it. If there are no type annotations, would that affect type stability, or would an error be caused [here](https://github.com/JuliaStats/DataFrames.jl/pull/1143/files#diff-15c11ad91975e38974d256d28f21db51R465) in any case?

Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: The annotation is only useful to ensure an error is printed immediately rather than later when the code attempts doing something which only works for symbols. In general it's better to strictly type arguments when it makes sense for clarity, and it makes the documentation clearer.

File: src/abstractdataframe/reshape.jl [hunk 8]
```diff
@@ -18,11 +18,16 @@ Stacks a DataFrame; convert from a wide to long format
 
 
 ```julia
-stack(df::AbstractDataFrame, measure_vars, id_vars)
-stack(df::AbstractDataFrame, measure_vars)
-stack(df::AbstractDataFrame)
-melt(df::AbstractDataFrame, id_vars, measure_vars)
-melt(df::AbstractDataFrame, id_vars)
+stack(df::AbstractDataFrame, measure_vars, id_vars;
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Please write `[measure_vars], [id_vars]` and get rid of the second definition. Same for `melt`. This is getting too verbose for my taste.

File: src/abstractdataframe/reshape.jl (hunk 3 above)
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: Since @andreasnoack's is the only one to have commented and he supports `variable_name`, could you change the (positional and keyword) arguments? Then it should be good to merge. Thanks.

File: src/abstractdataframe/reshape.jl (hunk 8 above)
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: i.e. condense the definition to a single one with optional measure/id vars?

File: src/abstractdataframe/reshape.jl (hunk 7 above)
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: OK, I shall add these annotations to all changed function declarations.

File: src/abstractdataframe/reshape.jl (hunk 3 above)
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: Whoops, github hid this comment. Fixed.

File: src/abstractdataframe/reshape.jl (hunk 5 above)
Comment (kdmurray91) [inline src/abstractdataframe/reshape.jl]: @nalimilan I believe this was fixed, could you double check that your requirement was met?

**Comment by kdmurray91** [comment]
@nalimilan Think it's good to go now.

File: .gitignore [hunk 9]
```diff
@@ -5,3 +5,5 @@ test/data/bigrams.tsv
 docs/build/
 docs/site/
 .DS_Store
+# Vim/vi
```
Comment (nalimilan) [inline .gitignore]: Unrelated, please open another PR if you want to have this merged.

**Comment by nalimilan** [comment]
Thanks, looks good except for the unrelated addition to `.gitignore`.

**Comment by nalimilan** [comment]
> EDIT: Is documenation some place other the docstrings required/wanted?

AFAIK no other changes are needed (only when adding new functions).

File: .gitignore (hunk 9 above)
Comment (kdmurray91) [inline .gitignore]: Ah crap, that was accidental, due to `git add .`. Reveted

**Comment by nalimilan** [comment]
Thanks!

File: src/abstractdataframe/reshape.jl [hunk 10]
```diff
@@ -65,49 +70,75 @@ d1 = DataFrame(a = repeat([1:3;], inner = [4]),
...
+
+function stack(df::AbstractDataFrame, measure_vars = numeric_vars(df);
+               variable_name::Symbol=:variable, value_name::Symbol=:value)
     mv_inds = index(df)[measure_vars]
-    stack(df, mv_inds, _setdiff(1:ncol(df), mv_inds))
+    stack(df, mv_inds, _setdiff(1:ncol(df), mv_inds);
+          variable_name=variable_name, value_name=value_name)
 end
 
 """
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: If you feel like it, it would be nice to remove this docstring in favor of the one from above, splitting `stack` and `melt`. Another option would be to use exactly the same docstring for both, I think some doc tricks allow doing this.
